### PR TITLE
Fix bugs in S.N.Sockets on *nix.

### PR DESCRIPTION
**Pull request by pgavlin**
- Address most remaining issues with S.N.Sockets on Linux
  - There are some differences in dual-mode semantics and error codes that are as of yet unresolved.
- Address a few bugs on OS X, mostly w.r.t. kqueue event registration.

### CODE REVIEW

**Comment by pgavlin** [comment]
@stephentoub PTAL

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 1]
```diff
@@ -346,26 +346,30 @@ public SocketAsyncContext(int fileDescriptor, SocketAsyncEngine engine)
...
-            Debug.Assert(!_handle.IsAllocated);
-            Debug.Assert(_registeredEvents == SocketAsyncEvents.None);
+            Debug.Assert(!_handle.IsAllocated || _registeredEvents != SocketAsyncEvents.None);
+            Debug.Assert((_registeredEvents & events) == SocketAsyncEvents.None);
 
-            _handle = GCHandle.Alloc(this, GCHandleType.Normal);
-
-            const SocketAsyncEvents Events = SocketAsyncEvents.Read | SocketAsyncEvents.Write;
+            if (_registeredEvents == SocketAsyncEvents.None)
+            {
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:355]: Just to be safe, consider adding an assert that `!_handle.IsAllocated`

Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:355]: Good idea. Done.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 2]
```diff
@@ -377,8 +381,15 @@ private void UnregisterRead()
 
             Interop.Error errorCode;
             bool unregistered = _engine.TryRegister(_fileDescriptor, _registeredEvents, events, _handle, out errorCode);
-            Debug.Assert(unregistered, string.Format("UnregisterRead failed: {0}", errorCode));
-            _registeredEvents = events;
+
+            if (unregistered || errorCode == Interop.Error.EBADF)
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: In what situations will we get EBADF back?

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs [hunk 3]
```diff
@@ -400,17 +411,20 @@ private void Unregister()
...
                 _handle.Free();
             }
+            else
+            {
+                Debug.Fail(string.Format("Unregister failed: {0}", errorCode));
+            }
         }
 
-        public void Close()
+        private void ProcessClose()
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: Nit: Consider a different name, like CloseCore, CloseInner, ClosePrivate, something like that.  Took me a minute to realize that this was about processing the close, not about closing the process ;)

Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: Done.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs (hunk 2 above)
Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs]: We may get this (or a few other errors) in the case that the file descriptor has already been closed.

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs [hunk 4]
```diff
@@ -72,9 +72,23 @@ public unsafe void EventLoop()
...
+                    // handle the HUP condition.
+                    if ((evts & Interop.libc.EPOLLHUP) != 0)
+                    {
+                        evts = (evts & ~Interop.libc.EPOLLHUP) | Interop.libc.EPOLLIN | Interop.libc.EPOLLOUT;
+                    }
+
                     var handle = (GCHandle)events[i].data;
                     var context = (SocketAsyncContext)handle.Target;
-                    context.HandleEvents(GetSocketAsyncEvents(events[i].events));
+                    if (context != null)
```
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: What does it mean if context is null?  Is it something we should assert for, or is it expected in some cases?  Seems erroneous that a GCHandle's target would be null.

Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: Consider the following sequence of events:
- `epoll_wait` or `kevent64` returns on thread 1 with an event for `SocketAsyncContext` A
- User code calls `Socket.Dispose` on thread 2, which ends up calling `SocketAsyncContext.Close()` on A, unregistering the underyling file descriptor and freeing the GC handle
- The event processing loop on thread 1 attempts to process the event for `SocketAsyncContext` A after its `GCHandle` has been freed, so `Target` is `null`.

**Comment by stephentoub** [comment]
LGTM

File: src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs (hunk 4 above)
Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: Oh, we're concerned about potentially using a handle after it's been freed?  GCHandles are very susceptible to having their ids be recycled, so if we're concerned about use-after-free, seems like we're susceptible here to getting the SocketAsyncContext for an unrelated operation...?

Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: Interesting--yes, that's something that we'll need to concern ourselves with. I'll put together a fix.

Comment (pgavlin) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: On second thought, handling this is going to be thornier than it seemed. I'll add a TODO for now.

Comment (stephentoub) [inline src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngineBackend.Linux.cs:88]: Ok.  Also keep in mind that Target can throw an exception in the case where the handle was freed: https://msdn.microsoft.com/en-us/library/vstudio/system.runtime.interopservices.gchandle.target(v=vs.100).aspx

File: src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs [hunk 5]
```diff
@@ -114,6 +114,8 @@ public Socket(AddressFamily addressFamily, SocketType socketType, ProtocolType p
             _socketType = socketType;
             _protocolType = protocolType;
 
+            Debug.Assert(addressFamily != AddressFamily.InterNetworkV6 || !DualMode);
```
Comment (CIPop) [inline src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:117]: @pgavlin DualMode is actually making a WinSock call on Windows. This `Debug.Assert` might come with odd side-effects. Could we instead track a `DEBUG` only variable that checks if `DualMode` has been set to true?

A `TODO` should suffice for now with the explanation to take care of this later.

**Comment by CIPop** [comment]
A single concern in the common/Win code otherwise LGTM.

**Pull request merged**